pkg/controller/migration/convert/network.go

```
@@ -0,0 +1,165 @@
+		// use the hardcoded value.
+		mtu := int32(c.calicoCNIConfig.MTU)
+		install.Spec.CalicoNetwork.MTU = &mtu
+	}
+
+	// CNI IPAM pools
+	if len(c.calicoCNIConfig.IPAM.IPv4Pools) != 0 {
+		return ErrIncompatibleCluster{"cni ipam ranges not suported"}
+	}
+
+	// Other CNI features
+	if c.calicoCNIConfig.FeatureControl.FloatingIPs {
+		return ErrIncompatibleCluster{"floating IPs not supported"}
+	}
+	if c.calicoCNIConfig.FeatureControl.IPAddrsNoIpam {
+		return ErrIncompatibleCluster{"IpAddrsNoIpam not supported"}
+	}
+
+	return nil
+}
+
+// getAutoDetection auto-detects the IP and Network using the requested
+// detection method.
+func getAutoDetection(method string) (operatorv1.NodeAddressAutodetection, error) {
+	const (
+		AUTODETECTION_METHOD_FIRST          = "first-found"
+		AUTODETECTION_METHOD_CAN_REACH      = "can-reach="
+		AUTODETECTION_METHOD_INTERFACE      = "interface="
+		AUTODETECTION_METHOD_SKIP_INTERFACE = "skip-interface="
+	)
+
+	// first-found
+	if method == "" || method == AUTODETECTION_METHOD_FIRST {
+		var t = true
+		return operatorv1.NodeAddressAutodetection{FirstFound: &t}, nil
+	}
+
+	// interface
+	if strings.HasPrefix(method, AUTODETECTION_METHOD_INTERFACE) {
+		ifStr := strings.TrimPrefix(method, AUTODETECTION_METHOD_INTERFACE)
+		return operatorv1.NodeAddressAutodetection{Interface: ifStr}, nil
+	}
+
+	// can-reach
+	if strings.HasPrefix(method, AUTODETECTION_METHOD_CAN_REACH) {
+		dest := strings.TrimPrefix(method, AUTODETECTION_METHOD_CAN_REACH)
+		return operatorv1.NodeAddressAutodetection{CanReach: dest}, nil
+	}
+
+	// skip-interface
+	if strings.HasPrefix(method, AUTODETECTION_METHOD_SKIP_INTERFACE) {
+		ifStr := strings.TrimPrefix(method, AUTODETECTION_METHOD_SKIP_INTERFACE)
+		return operatorv1.NodeAddressAutodetection{SkipInterface: ifStr}, nil
+	}
+
+	return operatorv1.NodeAddressAutodetection{}, errors.New("unrecognized option for AUTODETECTION_METHOD_SKIP_INTERFACE: " + method)
```

AUTODETECTION_METHOD_SKIP_INTERFACE isn't the right thing to report error on. I wonder if this function should return an error that the higher level function should add context to (because this function can be used for IPv6 or IPv4.

---

Good point. I adjusted the message and wrapped that context from the caller :+1: